src/main/java/org/opensearch/ad/task/ADBatchTaskRunner.java

```diff
@@ -306,28 +311,96 @@ public void getTopEntities(ADTask adTask, ActionListener<String> internalHCListe
+                    dataStartTime,
+                    dataStartTime + detectorInterval,
+                    internalHCListener
+                );
+            } else {
+                searchTopEntitiesForSingleCategoryHC(
+                    adTask,
+                    priorityTracker,
+                    dataEndTime,
+                    Math.max((dataEndTime - dataStartTime) / MAX_TOP_ENTITY_SEARCH_BUCKETS, detectorInterval),
+                    dataStartTime,
+                    dataStartTime + detectorInterval,
+                    internalHCListener
+                );
+            }
+        }, internalHCListener);
+    }
+
+    private void searchTopEntitiesForMultiCategoryHC(
+        ADTask adTask,
+        PriorityTracker priorityTracker,
+        long detectionEndTime,
+        long bucketInterval,
+        long dataStartTime,
+        long dataEndTime,
+        ActionListener<String> internalHCListener
+    ) {
+        checkIfADTaskCancelled(adTask.getTaskId());
+        ActionListener<List<Entity>> topEntitiesListener = ActionListener.wrap(topEntities -> {
+            topEntities
+                .forEach(entity -> priorityTracker.updatePriority(adTaskManager.convertEntityToString(entity, adTask.getDetector())));
+
+            if (dataEndTime < detectionEndTime) {
+                searchTopEntitiesForMultiCategoryHC(
+                    adTask,
+                    priorityTracker,
+                    detectionEndTime,
+                    bucketInterval,
+                    dataEndTime,
+                    dataEndTime + bucketInterval,
+                    internalHCListener
+                );
+            } else {
+                logger.debug("finish searching top entities at " + System.currentTimeMillis());
+                // remove HC detector level task from cache
+                adTaskCacheManager.remove(adTask.getTaskId());
+                List<String> topNEntities = priorityTracker.getTopNEntities(maxTopEntitiesPerHcDetector);
+                if (topNEntities.size() == 0) {
+                    logger.error("There is no entity found for detector " + adTask.getDetectorId());
+                    internalHCListener.onFailure(new ResourceNotFoundException(adTask.getDetectorId(), "No entity found"));
+                    return;
+                }
+                adTaskCacheManager.addPendingEntities(adTask.getDetectorId(), topNEntities);
+                adTaskCacheManager.setTopEntityCount(adTask.getDetectorId(), topNEntities.size());
+                internalHCListener.onResponse("Get top entities done");
+            }
+        }, e -> {
+            logger.error("Failed to get top entities for detector " + adTask.getDetectorId(), e);
+            internalHCListener.onFailure(e);
+        });
+
+        int minimumDocCount = Math.min((int) (bucketInterval / adTask.getDetector().getDetectorIntervalInMilliseconds()) / 2, 1);
```

Comment:
Why have a `Math.min` with a hardcoded `1` value? Wouldn't we always want the min count to be at least 1? This would allow to be 0 if the `bucketInterval` is small enough

Reply:
Good catch, should be Math.max.

`bucketInterval` should not be smaller than `detectorInterval`, check line 338 of this class
> Math.max((dataEndTime - dataStartTime) / MAX_TOP_ENTITY_SEARCH_BUCKETS, detectorInterval),

Reply:
> Wouldn't we always want the min count to be at least 1?

If the bucket interval is very big, for example it includes 1000 detector intervals, then we should not include entities which just has one doc in this bucket interval. From Kaituo, the priority tracker will refresh priority based on current bucket's top entities occurrence or not, ignoring their doc count. So we should not include such "small" entities in our top entities result. And it's hard to find a one-size-fits-all threshold for entity doc count, just use 1/2 interval counts as some easy threshold.

Reply:
Got it. `Math.max` makes sense here to raise the threshold, related to # intervals per bucket.